### PR TITLE
Check missed events before start

### LINKED ISSUE #5151: Events based cache could lose events at startup
When the server starts up there could be a skipped event that hasn't resolved. So upon restart we have to get a list of all the events and look for any gaps. This leads to the below issue:

```
Server 1 starts event 20 but doesnt finish
Server 2 creates event 21
Events < 20 get pruned
Server 3 starts up at event 21
Event 20 completes
```
In the scenario event 20 will never get processed by server 3.

* **Version**: 1.9.5
* **Platform**: all
* **Subsystem**: server events based cache

### PR DESCRIPTION
**Pull request by faisal-memon**
**Pull Request check list**

- [ ] Commit conforms to CONTRIBUTING.md?
- [x] Proper tests/regressions included?
- [ ] Documentation updated?

**Affected functionality**
Events based cache

**Description of change**
- Adds a new function `missedStartupEvents()` that checks for events below the first event we received.
- This runs evertime the cache is updated up till 24 hours, the transaction timeout time.
- Any events are immediately added to the cache.

**Which issue this PR fixes**
fixes #5151

### CODE REVIEW

**Comment by edwbuck** [comment]
@azdagron This is an important PR, being the one that closes 5151 and basically fixes the last of the known bugs in the database event processing.  If you are back from your travels, a quick look today might help us move it faster when we have Tuseday's meeting.

**Comment by edwbuck** [comment]
@amartinezfayo Would you please add this to the 1.10.1 release as was decided in the mainter's meeting on 7/16/2024?  This is a critical bug fix.

**Comment by azdagron** [comment]
I'd suggest a high-level explanation of the algorithm used to detect missing events and event gaps somewhere in the code. Things are a bit hard to follow and digging up the GH issue isn't convenient for folks trying to understand the code.

File: pkg/server/datastore/sqlstore/sqlstore.go [hunk 1]
```diff
@@ -4096,8 +4106,21 @@ func deleteRegistrationEntryEvent(tx *gorm.DB, eventID uint) error {
 
 func listRegistrationEntriesEvents(tx *gorm.DB, req *datastore.ListRegistrationEntriesEventsRequest) (*datastore.ListRegistrationEntriesEventsResponse, error) {
 	var events []RegisteredEntryEvent
-	if err := tx.Find(&events, "id > ?", req.GreaterThanEventID).Order("id asc").Error; err != nil {
-		return nil, sqlError.Wrap(err)
+
+	if req.GreaterThanEventID != 0 && req.LessThanEventID != 0 {
```
[the remark this answers is not in the document]
Comment (faisal-memon) [reply pkg/server/datastore/sqlstore/sqlstore.go]: done

File: pkg/server/endpoints/authorized_entryfetcher_attested_nodes.go [hunk 2]
```diff
@@ -23,30 +23,36 @@ type attestedNodes struct {
...
 	for _, event := range resp.Events {
-		if receivedFirstEvent && event.EventID != lastEventID+1 {
+		if !firstEventTime.IsZero() && event.EventID != lastEventID+1 {
 			for i := lastEventID + 1; i < event.EventID; i++ {
 				missedEvents[i] = clk.Now()
 			}
+		} else if firstEventTime.IsZero() {
+			firstEventID = event.EventID
+			firstEventTime = clk.Now()
 		}
```
[the remark this answers is not in the document]
Comment (faisal-memon) [reply pkg/server/endpoints/authorized_entryfetcher_attested_nodes.go:54]: i think so

**Review by azdagron — APPROVED** [review]
